### PR TITLE
Fixing held locks

### PR DESCRIPTION
**Pull request by hkaiser**
This should fix the current locking problems. 

@K-ballo could you please review the changes?

### CODE REVIEW

File: hpx/lcos/detail/future_data.hpp [hunk 1]
```diff
@@ -300,8 +301,11 @@ namespace detail
             wait(ec);
             if (ec) return NULL;
 
+            boost::unique_lock<mutex_type> l(mtx_);
```
Comment (K-ballo) [inline hpx/lcos/detail/future_data.hpp]: This lock is unnecessary. Once a future has been made ready, which is a postcondition of `wait`, either:

- there is only one writer (`future`), or
- there are multiple readers only (`shared_future`, lock hurts concurrency)

File: hpx/lcos/detail/future_data.hpp [hunk 2]
```diff
@@ -489,10 +496,20 @@ namespace detail
...
+
+            {
+                boost::unique_lock<mutex_type> l(this->mtx_);
+
+                std::swap(state_, s);
+                on_completed_ = completed_callback_type();
+
+                // unlock for the destruction of embedded data
+                // FIXME: is that correct?
+            }
```
Comment (K-ballo) [inline hpx/lcos/detail/future_data.hpp]: This change is unnecessary. A `reset` has to be non-observable, or futures would be unable to maintain their invariants. It can only be performed by a single writer with no readers.

File: hpx/lcos/detail/future_data.hpp [hunk 3]
```diff
@@ -423,6 +428,7 @@ namespace detail
 
             // check whether the data has already been set
             if (is_ready_locked()) {
+                l.unlock();
```
Comment (K-ballo) [inline hpx/lcos/detail/future_data.hpp]: Does `HPX_THROWS` suspend? Is that the origin of the held lock exceptions?

File: hpx/lcos/detail/future_data.hpp [hunk 4]
```diff
@@ -388,6 +392,7 @@ namespace detail
...
```
Comment (K-ballo) [inline hpx/lcos/detail/future_data.hpp]: This code can only be executed on a precondition violation, which would point to bugs elsewhere.

File: hpx/lcos/detail/future_data.hpp (hunk 3 above)
Comment (hkaiser) [inline hpx/lcos/detail/future_data.hpp]: THROWS might suspend as the stack is collected on a different HPX thread under certain circumstances, this has nothing to do with the problem at hand, it's just a fly-by.

File: hpx/lcos/detail/future_data.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/lcos/detail/future_data.hpp]: Ok, but the we don't need the lock in the first place. This lock was causing the issues when reset was invoked from the destructor.

File: hpx/lcos/detail/future_data.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/lcos/detail/future_data.hpp]: Ok, let's remove it again. I was not sure...

File: hpx/lcos/detail/future_data.hpp (hunk 2 above)
Comment (K-ballo) [inline hpx/lcos/detail/future_data.hpp]: Nod

File: hpx/lcos/detail/future_data.hpp (hunk 4 above)
Comment (hkaiser) [inline hpx/lcos/detail/future_data.hpp]: Agreed, it's an unrelated change.
